File: OrangeJuice.Server.Core/Services/AzureProductProvider.cs
using System;
using System.Threading.Tasks;

using Ab;
using Ab.Amazon.Data;
using Ab.Azure;
using Ab.Azure.Configuration;

namespace OrangeJuice.Server.Services
{
	public sealed class AzureProductProvider : IAzureProductProvider
	{
		#region Fields
		private readonly AzureOptions _azureOptions;
		private readonly IAzureClient _client;
		private readonly IConverter<string, Product> _converter;
		#endregion

		#region Ctor
		public AzureProductProvider(AzureOptions azureOptions, IAzureClient client, IConverter<string, Product> converter)
		{
			_azureOptions = azureOptions;
			_client = client;
			_converter = converter;
		}
		#endregion

		#region IAzureProductProvider members
		public async Task<Product> Get(Guid productId)
		{
			string content = await _client.GetBlobFromContainer(_azureOptions.ProductsContainer, productId.ToString());
			return content != null ? _converter.Convert(content) : null;
		}

		public Task Save(Product descriptor)
		{
			string content = _converter.ConvertBack(descriptor);
			return _client.PutBlobToContainer(_azureOptions.ProductsContainer, descriptor.ProductId.ToString(), content);
		}

		public Task<Uri> GetUrl(Guid productId)
		{
			return _client.GetBlobUrl(_azureOptions.ProductsContainer, productId.ToString());
		}
		#endregion
	}
}

File: OrangeJuice.Server.Core.FSharp/Services/AzureProductProvider.fs
namespace OrangeJuice.Server.FSharp.Services

open System
open System.Threading.Tasks

open OrangeJuice.Server
open OrangeJuice.Server.Configuration
open OrangeJuice.Server.Data.Models
open OrangeJuice.Server.Services

type AzureProductProvider(azureOptions : AzureOptions, client : IAzureClient, converter : IConverter<string, ProductDescriptor>) =
    interface IAzureProductProvider with
        member this.Get(productId : Guid) : Task<ProductDescriptor> =
            let task = async {
                let! content = client.GetBlobFromContainer(azureOptions.ProductsContainer, productId.ToString()) |> Async.AwaitTask
                return if content <> null 
                    then converter.Convert(content)
                    else null
            }
            task |> Async.StartAsTask

        member this.Save(descriptor : ProductDescriptor) : Task =
            let content = converter.ConvertBack(descriptor)
            client.PutBlobToContainer(azureOptions.ProductsContainer, descriptor.ProductId.ToString(), content)
        
        member this.GetUrl(productId : Guid) : Task<Uri> =
            client.GetBlobUrl(azureOptions.ProductsContainer, productId.ToString())
